Replacing FILENAME_MAX in runtime-directory

rchinmay opened the pull request:
Replaces char arrays in runtime directory, which used FILENAME_MAX macro for its size, to dynamic memory allocation using ``malloc`` and ``free`` functions.

rchinmay commented:
Resolves #8757 for runtime directory. 
@bradcray @daviditen Please review.
Thanks,
R. Chinmay

bradcray commented:
Hi @rchinmay — The failing check here is due to the direct calls to malloc() and free().  In practice in the runtime, we route all memory routines through the `chpl_mem_*` routines defined in `runtime/include/chpl-mem.h` (e.g., `chpl_mem_alloc()`, `chpl_mem_free()`, etc., so these malloc()/free() calls should be converted over to that interface.

Generally, these routines take in filenames and line numbers indicating the source of the event in terms of the user's source code.  However, if you `git grep` for other uses of them within `runtime/src/launch/*`, you'll find that we pass in dummy values like -1 and 0 in the runtime where there isn't an obvious user source location to pass in.

rchinmay commented:
@bradcray I have used ``CHPL_RT_MD_UNKNOWN`` in the description string argument of ``chpl_mem_allocMany`` in most of the cases. I didn't find any good description argument in ``runtime/include/chpl-mem-desc.h`` which matches in this allocation to file-names. Should I add any argument in the ``MEMDESCS`` for this purpose? If so, please suggest its name. If some argument already exists for this purpose, please suggest it.
Thanks,
R. Chinmay

bradcray commented:
@rchinmay: Good question, I had forgotten about that enumeration.  If it's not difficult, I like the idea of creating a new CHPL_RT_MD_FILENAME entry with description "filename string" or "filename buffer".  And organizationally, I might insert it just after the COMMAND_BUFFER entry since that seems to be the other description that's used heavily within the launchers.

rchinmay commented:
@bradcray I have added ``CHPL_RT_MD_FILENAME`` entry in the enumeration. Also, I have replaced CHPL_RT_MD_UNKNOWN in allocating memory to filenames with ``CHPL_RT_MD_FILENAME``. Please review.
Thanks,
R. Chinmay

bradcray commented:
@rchinmay : Thanks!  I didn't get to this today, but will try again next week.

bradcray commented on runtime/src/launch/pbs-aprun/launch-pbs-aprun.c:
```diff
@@ -393,10 +397,13 @@ static void genQsubScript(int argc, char *argv[], int numLocales) {
+      char* msg=(char *)chpl_mem_allocMany((strlen(expectFilename) + 
+                                            strlen(strerror(errno)) + 36), 
+                        sizeof(char), CHPL_RT_MD_FILENAME, -1, 0);
+      snprintf(msg, FILENAME_MAX + 45, "Error removing temporary file '%s': %s",
```
@rchinmay: Sorry that it's taken me awhile to get to this.  I have three main comments today, one important, and two related to making the code more readable or contain fewer magic numbers.  The first is the most important, of course, and the place to focus most.

I may be missing something obvious or having a brain freeze, but shouldn't the second argument to the `snprintf()` relate to the size of the `msg` buffer to ensure that we don't overflow it?  E.g., I would've expected a pattern like:

```c
int msgLen = strlen(expectFilename) + strlen(strerror(errno)) + 36;
char* msg = (char*)chpl_mem_allocMany(msgLen, sizeof(char), ...);
snprintf(msg, msgLen, "...");
```

While here, it might also be nice to tuck the magic number '35' into a variable to at least comment on where it comes from.  E.g.,

```c
int extraChars = 35;  // characters reserved for "Error removing temporary file '': "
```

Though it's imprecise, I could even imagine doing:

```c
const char* format = "Error removing temporary file '%s': %s";
int msgLen = strlen(expectFilename) + strlen(strerror(errno)) + strlen(format);
snprintf(msg, msgLen, format, ...);
```

This is inaccurate in that it's counting the `%s` strings as part of the string length, but I believe this will only overcount (and only by a few characters?), so it seems like it might be worthwhile to avoid the fragility of hardcoding the integer?  (You should definitely double-check my thinking, though, to make sure there isn't a problem I'm not seeing).

Let me know what you think.

rchinmay replied:
@bradcray 
>second argument to the snprintf() relate to the size of the msg buffer to ensure that we don't overflow it

Yes. Sorry about my mistake here. I will correct this. I might have missed the 2nd argument in ``snprintf()`` while changing this. 
Also I think using ``char* format`` would be better, because if the error message would be changed in future, then explicitly changing the integer (``extraChars``) would not be required and thus would be more prone to changes in such messages in future. Please tell me if I am wrong anywhere.
Thanks,
R. Chinmay.

bradcray replied:
That matches my thinking, thanks.

rchinmay replied:
@bradcray Please review. I have made the changes.

bradcray commented on runtime/src/launch/pbs-gasnetrun_ibv/launch-pbs-gasnetrun_ibv.c:
```diff
@@ -170,6 +170,17 @@ static char* chpl_launch_create_command(int argc, char* argv[],
   mypid = 0;
 #endif
+  sysFilename=(char *)chpl_mem_allocMany((strlen(baseSysFilename) + 
+                                    snprintf(NULL, 0, "%d", (int)mypid) + 1), 
```
I had not seen this pattern before—clever.  Given that this same expression appears three times in a row, how about storing it into an `int` and then re-using that int in these three malloc calls.  I.e., something like:

```c
int mypidBuffLen = snprintf(NULL, 0, "%d", (int)mypid);
```

bradcray commented on runtime/src/launch/slurm-srun/launch-slurm-srun.c:
```diff
@@ -324,7 +338,13 @@ static char* chpl_launch_create_command(int argc, char* argv[],
+                                    strlen(argv[0]) + strlen(tempArg);
+      tmpStdoutFileNoFmt = (char *)chpl_mem_allocMany(tmpStdoutFileNoFmtLen, 
+                                sizeof(char), CHPL_RT_MD_FILENAME, -1, 0);
+      sprintf(tmpStdoutFileNoFmt, format, tmpDir, argv[0], tempArg);
```
In cases like this where you've computed and stored the buffer length, wouldn't it be safer / more future-proof to use `snprintf()` rather than `sprintf()` (to avoid buffer overrun in case someone changes one part of the malloc/snprintf pattern in the future and doesn't change the other part to match?).  Essentially, I'm curious whether there was a rationale for using `sprintf()` in some cases and `snprintf()` in others.

rchinmay replied:
@bradcray
> I'm curious whether there was a rationale for using sprintf() in some cases and snprintf() in others.

I kept the usage of ``snprintf()`` and ``sprintf()`` the same as in the code before changes (this PR). But, now since we have the lengths of the char arrays with us, it would be better (safer) to convert ``sprintf()`` to ``snprintf()``, if we have the lengths of the char array already stored. This would also be more robust to changes in future, rather than ``sprintf()``. I will make these changes.
Thanks,
R. Chinmay

bradcray replied:
Your rationale makes sense, but if you're willing to make that change while here, I think that'd be great, thank you!

bradcray commented:
I think this is getting close, thanks @rchinmay (and apologies again for my slow response... you probably regret getting me as the lead reviewer).

I wanted to tag @ronawho here to see if he had thoughts about what I should do to test this before merging (I suspect there's no easy answer given that it's distinct launchers; but maybe I could test them all locally using asan just to make sure they "failed to launch" and didn't have memory errors, and then left the stress testing for the nightlies?)

Also tagging @cassella since he was the impetus behind the original issue #8757.

ronawho commented:
> I wanted to tag @ronawho here to see if he had thoughts about what I should do to test this before merging (I suspect there's no easy answer given that it's distinct launchers; but maybe I could test them all locally using asan just to make sure they "failed to launch" and didn't have memory errors, and then left the stress testing for the nightlies?)

That's an interesting idea, but it'll probably fail in unhelpful ways today since some launchers call things like `sinfo` or `cnselect` to build the launch commands so you'll probably get early failures. I wonder if we could add a `--dry-run` option or something that would fake values for calls and then print out what command the launcher would use. Kinda like `-v`, but don't actually launch and don't require launcher utilities like `cnselect`/`sinfo`

For this PR, I think just ping me when you're ready for testing and I'll spot check a few launchers.

bradcray commented:
> some launchers call things like sinfo or cnselect to build the launch commands

Oh, good point.

> For this PR, I think just ping me when you're ready for testing and I'll spot check a few launchers.

That'd be very nice, thanks.  I'd also be curious whether you'd prefer to hold off on merging this PR until after the release to live with it for awhile, or whether you think between your spot-checking and nightly testing we'd probably be safe enough.

cassella commented:
Hrm.  In `launch-pbs-aprun.c`, I'm seeing the function `genQsubOptions()` sprintf()s into the global `expectFilename`:

```c
static char* genQsubOptions(char* genFilename, char* projectString, qsubVersion qsub, 
                            int32_t numLocales, int32_t numCoresPerLocale) {
...
  char *qsubFilename = expectFilename;
...

  if (generate_qsub_script) {
    pid_t mypid = debug ? 0 : getpid();
    sprintf(qsubFilename, "qsub.%s-%d", genFilename, (int) mypid);
 ```

which may be a longer string.

This is pretty gross.  If it were up to me, I'd make `expectFilename` local to `chpl_launch_create_argv()`, and give `genQsubOptions()` its own `char *` to manage.

(It's not straightforward since `genQsubOptions()` uses the buffer and returns it in one case, and returns it unmodified in the other.  I think this is all just so the file didn't need to have two static buffers in it.  If so, I think it would be cleaner if they didn't try to share like that.  I don't know how much of a mess it'll be to straighten that out.)

bradcray commented:
@cassella : Just to make sure I'm following, your last comment is unrelated to this specific effort (retiring the use of FILENAME_MAX), is that right?  (I understand that it's thematically related in that it relates to how launcher string buffers are allocated and populated).

Assuming so, I propose we consider that a pre-existing condition and not worry about it here.  If something about this PR makes the existing code less attractive than it was, then I'd worry about it here.

cassella commented:
@bradcray No, it is a new problem with this PR.

Currently, `genQsubOptions()` sprintf()s into the `expectFilename` buffer.  Its pattern is small and the buffer is `PATH_MAX`, so that's fine.

With the PR, it's still sprintf()ing into the same buffer, but now the buffer is sized precisely according to the batch filename and pid.   Depending on what `genQsubOptions()`'s `genFilename` is, it may be sprintfing something larger than the buffer.


The "if it were up to me" part is just what I'd do about it, versus (I guess?) sizing the single array to the max of the two uses' lengths.

ronawho commented:
> That'd be very nice, thanks. I'd also be curious whether you'd prefer to hold off on merging this PR until after the release to live with it for awhile, or whether you think between your spot-checking and nightly testing we'd probably be safe enough.

At this point I'd probably hold off until after the release. (I was hoping to get to this last week, but other release tasks have been taking up all my time)

bradcray commented:
@rchinmay : Now that we're through with the release, I wanted to ping you to see whether you're still interested in pursuing this effort?  I think there were a few minor clean-up items in my last review that it looked like you'd intended to address; and then there's the more fundamental issue that @cassella brings up above.

rchinmay commented:
@bradcray, Thanks for informing on the release. I saw @cassella 's comment and I am currently working on it to solve the problem. Along with that, I will also make the clean-ups suggested by you. 
Thanks, 
R. Chinmay.

bradcray commented:
Great, thanks for the update!

bradcray commented:
@rchinmay : I'm looking through PRs that I'm associated with today and wanted to check in on this one.  Are you still working on it?  Thanks.

bradcray commented:
I'm going to park this until someone has a chance to get back to it.

riftEmber commented:
Picking this work back up in https://github.com/chapel-lang/chapel/pull/26381, which incorporates the commits from this PR branch.
